### DISCUSSION TITLE
xUnit to TUnit - How to implement multiple fixtures in a single CollectionDefinition?

### DISCUSSION THREAD

**Discussion by richardoSGSI** [thomhurst/TUnit#926] [Q&A] [answered]
Hi all,

I'm trying to convert a class like this:

```c#
using Tests.Common.Fixtures;
using Xunit;

namespace MyTests.Tests
{
  [CollectionDefinition("MyTests.Tests fixture collection")]
  public class FixtureCollection : ICollectionFixture<MyFixture1>, ICollectionFixture<MyFixture2>, ICollectionFixture<MyFixture3>
  {
    // This class has no code, and is never created. Its purpose is simply
    // to be the place to apply [CollectionDefinition] and all the
    // ICollectionFixture<> interfaces.
  }
}
```

and then use it like:

```c#
namespace MyTests.Tests
{
  [Collection("MyTests.Tests fixture collection")]
  public class Tests
  {
    public Tests (MyFixture1 myFixture1, MyFixture2 myFixture2, MyFixture3 myFixture3)
    {
	...
    }
  }
}
```

I saw that it's possible to do a single fixture via [this thread](https://github.com/thomhurst/TUnit/discussions/601):
```c#
[ClassDataSource<MyFixture>(Shared = SharedType.Globally)]
```

I also saw there were things like [DataSourceGeneratorAttribute](https://thomhurst.github.io/TUnit/docs/tutorial-extras/data-source-generators) and [ClassConstructorAttribute](https://thomhurst.github.io/TUnit/docs/tutorial-extras/class-constructors), but I'm not sure what would be appropriate

Any ideas? Thanks!

EDIT:
And as a follow-up question, if I used the `DataSourceGeneratorAttribute`, is it possible to set the scope for how I want it to be shared? Like how we're able to do so on the `ClassDataSource` attribute?

**Comment by thomhurst** [comment, marked as answer]
Hey @richardoSGSI  - I've just pushed v0.1.1063 which can support injecting up to 5 generic arguments via ClassDataSource.

So you could do:

```
namespace MyTests.Tests
{
  [ClassDataSource<MyFixture1, MyFixture2, MyFixture3>(Shared = SharedType.Globally)]
  public class Tests
  {
    public Tests (MyFixture1 myFixture1, MyFixture2 myFixture2, MyFixture3 myFixture3)
    {
	...
    }
  }
}

```

Regarding using the `DataSourceGeneratorAttribute` - You'd have to implement it yourself. `ClassDataSource` is actually built on top of `DataSourceGeneratorAttribute` - And simply uses the Shared argument to store already created instances in dictionaries etc.

So it's possible, but you'd have to implement it yourself. You can look at the source for `ClassDataSource` yourself to see how I've done it.

**Comment by thomhurst** [comment]
Just as a follow up, did injecting multiple fixtures via `ClassDataSource` resolve your problem?
